### ISSUE TITLE
sccache support

### ISSUE THREAD

**Issue by Trass3r** [mesonbuild/meson#5428]
Everything seems to be hardcoded to ccache.
https://github.com/mozilla/sccache also works on windows.

**Comment by dcbaker** [comment]
looking at their readme it looks like a "works for firefox, probably not for anything else" sort of project.

We certainly can generalize the ccache functionality (and it's probably worth doing) there's also [clcache](https://github.com/frerich/clcache), which is probably worth supporting if possible.

**Comment by jpakkane** [comment]
Note that you can do it manually with something like:

```
CC="ccache cl" CXX="ccache cl" meson <your options here>
```

**Comment by Trass3r** [comment]
> looking at their readme it looks like a "works for firefox, probably not for anything else" sort of project.

I guess you're refering to https://github.com/mozilla/sccache/blob/master/README.md#known-caveats.
Well if it can build Firefox it should be good enough for a lot of projects.

> We certainly can generalize the ccache functionality (and it's probably worth doing) there's also [clcache](https://github.com/frerich/clcache), which is probably worth supporting if possible.

👍

**Comment by Trass3r** [comment]
> CC="ccache cl" CXX="ccache cl" meson <your options here>

By the way is this limited to Ninja?
Looks like you can do that kind of thing in the VisualStudio project file too:
```xml
<PropertyGroup>
  <CLToolExe>clang-cl.exe</CLToolExe>
  <CLToolPath>C:\llvm\build\bin</CLToolPath>
</PropertyGroup>
```

**Comment by dcbaker** [comment]
Honestly most of us are Linux/BSD devs primarily, getting that kind of support added to the MSVC backend is on my list of things to do.

**Comment by pgallir** [comment]
> CC="ccache cl" CXX="ccache cl" meson <your options here>

just a silly question: would not be the same for sccache? i.e., 
`CC="sccache cl" CXX="sccache cl" meson <your options here>`

**Comment by pgallir** [comment]
On the other hand, to my understanding, Bazel provides [remote caching](https://docs.bazel.build/versions/master/remote-caching.html#run-bazel-using-the-remote-cache) using some [backends](https://docs.bazel.build/versions/master/remote-caching.html#setting-up-a-server-as-the-caches-backend), which is suitable feature for a build factory:  if builds can be safely reused on other machines, the overall compilation time might be incredibly faster.

Is there something like this ongoing for Meson project as well? Would [sccache ](https://github.com/pgallir/sccache) or [xcache](https://github.com/alibaba/xcache)  be a suitable backend to exploit cache sharing?

Though [distcc ](https://github.com/pgallir/distcc) or (even better, as it can schedule builds across nodes) [Icecream](https://github.com/pgallir/icecream) are valid options to distribute build jobs, but they lack of "memory", which might be empowered by sc/x-cache projects.

**Comment by nirbheek** [comment]
Would be nice to have comprehensive/pluggable support for compiler cache tools in Meson.

**Comment by philn** [comment]
I've tried `CC="sccache cc"` but it doesn't work, meson seems to insist on using ccache: `ccache: error: Recursive invocation (the name of the ccache binary must be "ccache")`

**Comment by philn** [comment]
Sorry for the noise, misconfiguration here. It works :)

**Comment by eli-schwartz** [comment]
This is now implemented in #9428 and coming soon to a meson near you. ;)

**Comment by pgallir** [comment]
cool :) 👍
